File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        
        stack<TreeNode*> st;
        TreeNode* curr = root;
        int cnt = 0;
        
        while(true){
            if(curr!=NULL){
                st.push(curr);
                curr = curr->left;
            }else{
                if(st.empty()) break;
                
                curr = st.top();
                st.pop();
                cnt++;
                
                if(cnt==k) return curr->val;
                curr = curr->right;
                
            }
        }
        
        return -1;
        
    }
};

File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def helper(self,root):
        if root:
            self.helper(root.left)
            self.k -=1
            if self.k  == 0:
                self.ans = root.val
                return
            
            self.helper(root.right)
            
        
        
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        
        # inorder of bst is sorted
        
        self.k = k
        self.ans = 0
        self.helper(root)
        return self.ans
        